src/tools/particles/Controls.tsx

```diff
@@ -0,0 +1,124 @@
+import * as React from 'react'
+
+import { Config } from '~/tools/particles/interfaces'
+
+interface ConfigWithBit extends Config {
+  dirty: boolean
+}
+
+export const Controls = (props: {
+  initialConfig: Config
+  updateGlobalConfig: (c: Config) => void
+}) => {
+  const [config, setConfig] = React.useState<ConfigWithBit>({
+    ...props.initialConfig,
+    dirty: false,
+  })
+
+  const updateConfig = (key: string, value: any) => {
+    setConfig({ ...config, [key]: value, dirty: true })
+  }
+
+  const publishConfig = () => {
+    props.updateGlobalConfig(config)
+    setConfig({ ...config, dirty: false })
+  }
+
+  const parse = (val: string) => {
+    if (val === '') return ''
```

Comment:
Should we maybe return `NaN` so the return types are consistent here?

Reply:
Specifically, this is here so that it _doesn't_ print `NaN` into the browser field, which looks unexpected